logendpoint.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2017  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "logendpoint.h"
+
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "mainloop.h"
+#include "util.h"
+
+static bool _logging_start_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_start_timeout();
+}
+
+static bool _alive_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_alive_timeout();
+}
+
+void LogEndpoint::_send_msg(const mavlink_message_t *msg, int target_sysid)
+{
+    uint8_t data[MAVLINK_MAX_PACKET_LEN];
+    struct buffer buffer { 0, data };
+
+    buffer.len = mavlink_msg_to_send_buffer(data, msg);
+    _mainloop->route_msg(&buffer, target_sysid, msg->sysid);
+
+    _stat.read.total++;
+    _stat.read.handled++;
+    _stat.read.handled_bytes += msg->len;
+    if (msg->magic == MAVLINK_STX)
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink2_header);
+    else
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink1_header);
+}
+
+char *LogEndpoint::_get_filename(const char *extension)
+{
+    char filename[PATH_MAX] = { };
+
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
```

Comment:
This is awesome - if it works.

I initially wrote the dronekit-la implementation doing this, but had to fall back to sequential numbering; the issue is that the real-time-clock on whatever computer you are running this on may not be initialised from an external source yet.  On Solo this was from the GPS passed through the flight control board 0 if you're inside you may never get a time update, and when it comes wackiness ensues.